### DISCUSSION TITLE
is it possible to `include` a markdown file in a liquid template?

### DISCUSSION THREAD

**Discussion by ahillio** [11ty/eleventy#2289] [Q&A]
I have a liquid template for displaying a collection. I want to do `{% include content.md %}` but can't find any solutions. I would absolutely settle for having to `{% include content_layout.liquid %}` and use that `content_layout.liquid` file to render my `content.md` file... or any other way to do this ...but I've tried that and the markdown content file does not get displayed.

This seems like it would be a fairly common thing to do, have a piece of content at the top of your collection and be able to edit that content in a markdown file instead of html inside the template.

Are there any blogs or docs which explain this? I've checked and websearched and the only thing I found was on stackoverflow saying "11ty doesn't work that way" :(

**Comment by pdehaan** [comment]
@ahillio If you're using Eleventy v1.0.0+, I'd suggest using the EleventyRenderPlugin plugin (specifically the `renderFile` shortcode): https://www.11ty.dev/docs/plugins/render/#renderfile

```md
<!-- ./src/_includes/content.md -->

## I am some content

I am a paragraph

- i am
- a list

> blockquote
```

```liquid
---
# src/index.liquid
title: "11ty-2289"
---

<h1>{{ title }}</h1>

{% renderFile "./src/_includes/content.md" %}

<footer>
  <p>&copy; 2099</p>
</footer>
```

```js
// .eleventy.js
const { EleventyRenderPlugin } = require("@11ty/eleventy");

module.exports = (eleventyConfig) => {
  eleventyConfig.addPlugin(EleventyRenderPlugin);

  return {
    dir: {
      input: "src",
      output: "www",
    }
  };
};
```

### OUTPUT

```html
<!-- ./src/_includes/content.md -->
<h2>I am some content</h2>
<p>I am a paragraph</p>
<ul>
  <li>i am</li>
  <li>a list</li>
</ul>
<blockquote>
 <p>blockquote</p>
</blockquote>
```

**Reply by pdehaan** [reply]
If you're NOT using Eleventy 1.0+ (and you should be, it's like, super neat), you could always create a Markdown renderer instance via markdown-it and create a custom filter.

```liquid
<!-- ./src/index.liquid -->
{% capture md_content %}
  {% render "content.md" %}
{% endcapture %}
{{ md_content | markdownify }}

{{ "**bold** and _italics_" | markdownify: true }}
{{ "**bold** and _italics_" | markdownify: false }}
```

```js
// .eleventy.js
const markdownIt = require("markdown-it");

module.exports = (eleventyConfig) => {
  const md = markdownIt({
    html: true,
    linkify: true,
    typographer: true,
  });

  eleventyConfig.addFilter("markdownify", function (str="", inline=false) {
    if (inline) {
      return md.renderInline(str);
    }
    return md.render(str);
  });

  return {
    dir: {
      input: "src",
      output: "www",
    }
  };
};
```

### OUTPUT
```html
<!-- ./src/_includes/content.md -->
<h2>I am some content</h2>
<p>I am a paragraph</p>
<ul>
  <li>i am</li>
  <li>a list</li>
</ul>
<blockquote>
  <p>blockquote</p>
</blockquote>

<strong>bold</strong> and <em>italics</em>
<p><strong>bold</strong> and <em>italics</em></p>
```

**Reply by pdehaan** [reply]
And I'm sorry, this is all impossibly confusing.
- LiquidJS added [`{% render %}`](https://liquidjs.com/tags/render.html) and deprecated [`{% include %}`](https://liquidjs.com/tags/include.html) tags in v9.2.0.
- Eleventy added a [EleventyRenderPlugin](https://www.11ty.dev/docs/plugins/render/) plugin in v1.0.0 which includes `{% renderTemplate %}` and `{% renderFile %}` shortcodes.
- markdown-it has both [`render()`](https://markdown-it.github.io/markdown-it/#MarkdownIt.render) and [`renderInline()`](https://markdown-it.github.io/markdown-it/#MarkdownIt.renderInline) methods (the latter is "similar to [MarkdownIt.render](https://markdown-it.github.io/markdown-it/#MarkdownIt.render) but for single paragraph content. Result will NOT be wrapped into `<p>` tags.").

It's ~~turtles~~ renders all the way down.

**Reply by ahillio** [reply]
Awesome! I'm looking into upgrading to 1.x right now. Is it easy to downgrade/revert 11ty back to my current working version if the breaking changes (I'm reading them at https://www.11ty.dev/blog/eleventy-one-point-oh/) are too troublesome?

**Reply by pdehaan** [reply]
I think most people say upgrading has been really painless, but it really depends on the complexity of the site.
Downgrading should be as easy as reverting the package.json version and reinstalling dependencies.

There is also a migration wizard plug-in at https://github.com/11ty/eleventy-upgrade-help to tell you what, if anything, might need to change.

I think the biggest issue so far with upgrading is some differences with strict filters in newer versions of liquidjs (v6 to v9). Comically, I think every time that has been a problem for somebody, it turns out that liquid is throwing errors instead of failing to render content, so if anything it’s finding bugs that were previously there.

plus if your site is in a public repo, I can clone it locally and help take a look.

**Reply by ahillio** [reply]
I'm getting
```
`TemplateContentRenderError` was thrown
[11ty] > Could not find render plugin file for the `renderFile` shortcode,
```
Does that mean 11ty isn't properly updated to 1.0.0? It seemed to be updated because I ran into https://github.com/11ty/eleventy/issues/1677 that error.

https://github.com/11ty/eleventy-upgrade-help hasn't given me any more error information (the readme doesn't give any usage instructions) I did `npm install` and then edited .eleventy.js as it said.

**Reply by pdehaan** [reply]
When you do a build, the logging should tell you what version is currently installed:

```sh
[11ty] Wrote 1 file in 0.04 seconds (v1.0.0)
```

Not much help if the build can't finish though, but might be buried somewhere in the output.
Otherwise, I usually resort to `console.log("Eleventy version:", require("@11ty/eleventy/package.json").version);` near the top of my .eleventy.js file or whatever.

A bit hard to troubleshoot that `renderFile` error, but if I remember my errors correctly while debugging, it was because the `{% renderFile "./src/_includes/content.md" %}` needed a full path to the file in question relative to the .eleventy.js/root. I think initially I was just trying to do "content.md" but it wasn't resolving to the _includes/ directory. Then I tried "_includes/content.md" but it didn't like that, so I had to put in the input dir as well.

**Reply by rdela** [reply]
@pdehaan thanks for this! Helped me just now. 

```nunjucks
{% renderFile "./_includes/about.md" %}
{#  works much better than #}
{% include "about.md" %}
```

Especially in `.njk` templates, where `include` was not wrapping paragraphs in `p` tags, for instance (`include` ***was*** wrapping paragraphs in `p` tags in `.md` pages though). 

Also very cool that… 

> The first argument to `renderFile` is a project root relative path to any template file.

…And not relative to the including page or tempate.
